File: backend/src/entropia/apps/api/routes/sharing.py
"""Explicit package-sharing API (GAP-17; Master Reference §6, §6.4).

Thin handlers over the sharing command/query layer — no SQL, policy or business
logic here. The grant/revoke mutations carry the package root ``row_version`` as
the ``If-Match "rv-N"`` ETag (the ``GET /library/{entity_id}`` detail returns it)
plus a fresh ``Idempotency-Key``, matching the sibling package Move-to-Trash
route. ``GET /library/{entity_id}/shares`` lists the active grantees and is
owner/Admin-only (a grantee never learns who else a resource is shared with).
``GET /library-shared-with-me`` is the grantee's inbox of shared packages (a
distinct literal path so it never shadows ``GET /library/{entity_id}``).
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Header, Query, Response
from pydantic import BaseModel

from entropia.application.commands import sharing as sharing_cmd
from entropia.application.queries import library as library_query
from entropia.application.queries import sharing as sharing_query
from entropia.apps.api.deps import RequestContext, request_context
from entropia.shared.concurrency import etag_for_row_version, row_version_from_if_match
from entropia.shared.pagination import PageParams

router = APIRouter(tags=["package-sharing"])


class SharePackageRequest(BaseModel):
    grantee_email: str


@router.post("/library/{entity_id}/shares", status_code=201)
async def share_package(
    entity_id: str,
    body: SharePackageRequest,
    ctx: RequestContext = Depends(request_context),
    if_match: str | None = Header(default=None, alias="If-Match"),
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
) -> dict[str, Any]:
    return await sharing_cmd.share_package(
        ctx.session,
        ctx.actor,
        entity_id=entity_id,
        grantee_email=body.grantee_email,
        expected_row_version=row_version_from_if_match(if_match),
        idempotency_key=idempotency_key,
    )


@router.get("/library/{entity_id}/shares")
async def list_package_shares(
    entity_id: str,
    response: Response,
    ctx: RequestContext = Depends(request_context),
) -> dict[str, Any]:
    result = await sharing_query.list_package_shares(ctx.session, ctx.actor, entity_id=entity_id)
    response.headers["ETag"] = etag_for_row_version(int(result["row_version"]))
    return result


@router.delete("/library/{entity_id}/shares/{share_id}")
async def revoke_package_share(
    entity_id: str,
    share_id: str,
    ctx: RequestContext = Depends(request_context),
    if_match: str | None = Header(default=None, alias="If-Match"),
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
) -> dict[str, Any]:
    return await sharing_cmd.revoke_package_share(
        ctx.session,
        ctx.actor,
        entity_id=entity_id,
        share_id=share_id,
        expected_row_version=row_version_from_if_match(if_match),
        idempotency_key=idempotency_key,
    )


@router.get("/library-shared-with-me")
async def list_shared_with_me(
    cursor: str | None = Query(default=None),
    limit: int = Query(default=20, ge=1, le=100),
    ctx: RequestContext = Depends(request_context),
) -> dict[str, Any]:
    return await library_query.list_shared_with_me(
        ctx.session, ctx.actor, PageParams(cursor=cursor, limit=limit)
    )

File: frontend/src/lib/sharing.ts
// Explicit package sharing (GAP-17; Master Reference §6, §6.4). The owner (or an
// Admin) grants a specific user explicit view/use access to a private package;
// the backend flips PRIVATE <-> EXPLICITLY_SHARED as the first/last grant is
// added/removed. Grant + revoke carry the package root row_version as the
// If-Match "rv-N" OCC token (the /library/{id} detail returns it) + a fresh
// Idempotency-Key — mirroring the sibling Move-to-Trash mutation. Listing the
// grantees is owner/Admin-only. "Shared with me" is the grantee's inbox. All
// read keys live under ["library"] (swept by resource.changed); mutations also
// sweep ["audit"].

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import type { QueryClient } from "@tanstack/react-query";

import { api, apiRequest } from "./apiClient";
import type { LibraryPage } from "./library";

// ---------------------------------------------------------------------------
// Wire types (mirror application/queries/sharing.py + commands/sharing.py)
// ---------------------------------------------------------------------------

export interface PackageShare {
  share_id: string;
  grantee_principal_id: string;
  grantee_email: string | null;
  grantee_display_name: string | null;
  granted_by_principal_id: string | null;
  created_at: string | null;
}

export interface PackageSharesResult {
  entity_id: string;
  visibility_scope: string;
  row_version: number;
  shares: PackageShare[];
}

export interface ShareResult {
  entity_id: string;
  share_id: string;
  grantee_principal_id: string;
  grantee_email: string | null;
  grantee_display_name: string | null;
  visibility_scope: string;
  active_share_count: number;
  row_version: number;
}

export interface RevokeShareResult {
  entity_id: string;
  share_id: string;
  revoked: boolean;
  visibility_scope: string;
  active_share_count: number;
  row_version: number;
}

// ---------------------------------------------------------------------------
// Read hooks (["library"] prefix)
// ---------------------------------------------------------------------------

// The active grantees of a package. Owner/Admin-only server-side (a grantee
// gets 403 SHARE_MANAGEMENT_FORBIDDEN, rendered verbatim); the client enables
// the query only where the server marks `can_share`, but that is never the
// authority — the server re-validates.
export function usePackageShares(entityId: string | null, enabled: boolean) {
  return useQuery({
    queryKey: ["library", "shares", entityId],
    queryFn: () =>
      api.get<PackageSharesResult>(
        `/library/${encodeURIComponent(entityId ?? "")}/shares`,
      ),
    enabled: entityId !== null && enabled,
  });
}

// The caller's inbox of packages explicitly shared WITH them (GAP-17).
export function useSharedPackages() {
  return useQuery({
    queryKey: ["library", "shared-with-me"],
    queryFn: () => api.get<LibraryPage>("/library-shared-with-me"),
  });
}

// ---------------------------------------------------------------------------
// Mutations — OCC via If-Match "rv-N" + a fresh Idempotency-Key per attempt.
// The client never pre-gates on permissions; the server renders 403/409/422
// verbatim.
// ---------------------------------------------------------------------------

function invalidateSharing(queryClient: QueryClient) {
  void queryClient.invalidateQueries({ queryKey: ["library"] });
  void queryClient.invalidateQueries({ queryKey: ["audit"] });
}

export function useSharePackage() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: { entityId: string; rowVersion: number; granteeEmail: string }) =>
      apiRequest<ShareResult>(`/library/${encodeURIComponent(input.entityId)}/shares`, {
        method: "POST",
        headers: {
          "If-Match": `"rv-${input.rowVersion}"`,
          "Idempotency-Key": crypto.randomUUID(),
        },
        body: { grantee_email: input.granteeEmail },
      }),
    onSuccess: () => invalidateSharing(queryClient),
  });
}

export function useRevokeShare() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: { entityId: string; shareId: string; rowVersion: number }) =>
      apiRequest<RevokeShareResult>(
        `/library/${encodeURIComponent(input.entityId)}/shares/${encodeURIComponent(input.shareId)}`,
        {
          method: "DELETE",
          headers: {
            "If-Match": `"rv-${input.rowVersion}"`,
            "Idempotency-Key": crypto.randomUUID(),
          },
        },
      ),
    onSuccess: () => invalidateSharing(queryClient),
  });
}
